tests/tests_e3/conftest.py

```diff
@@ -164,3 +175,87 @@ def pytest_sessionfinish(session, exitstatus):
         # pytest docs, which is the closest thing to having an actual Python error in
         # test code.
         session.exitstatus = 3
+
+
+class PypiSimulator:
+    PYPI_URL = "https://pypi.org"
+    MATCHER = regex_compile(f"{PYPI_URL}/pypi/(?P<package>.*)/json")
+    DATA_DIR = path_join(dirname(abspath(__file__)), "pypi_data")
+
+    def __init__(self, requests_mock: Any) -> None:
+        self.requests_mock = requests_mock
+        self.requests_mock.stop()
+        self.download_url_mocked = set()
+
+    def download_file(self, name: str, version: str, request: Any, context: Any) -> str:
+        if request.url not in self.download_url_mocked:
+            raise Exception("File url not mocked")
+
+        mkdir(name)
+
+        with open(path_join(name, "setup.py"), "w") as fd:
+            fd.write("from setuptools import setup, find_packages\n")
+            fd.write(f"setup(name='{name}',\n")
+            fd.write(f"      version='{version}',\n")
+            fd.write("       packages=find_packages())\n")
+        mkdir(os.path.join(name, name))
+        with open(os.path.join(name, name, "__init__.py"), "w") as fd:
+            fd.write(f"# This is package {name}")
+
+        pkg = Wheel.build(source_dir=name, dest_dir=".")
+        assert isfile(pkg.path)
+        with open(pkg.path, "rb") as f:
+            result = f.read()
+
+        context.status_code = 200
+        return result
+
+    def get_metadata(self, request: Any, context: Any) -> str:
+        m = self.MATCHER.match(request.url)
+        if not m:
+            raise ValueError("mocked pypi received an unexpected request")
+        package = m.group("package")
+        path = path_join(self.DATA_DIR, "json", f"{package}.json")
+        if not isfile(path):
+            context.status_code = 404
+            return {"message": "Not Found"}
+
+        try:
+            with open(path) as json_file:
+                result = json_loads(json_file.read())
+
+            if "releases" not in result:
+                raise Exception("Bad metadata")
+        except Exception:
```

Comment:
Here we lose the exception history. Maybe creating an `InvalidMetadataError` exception, and then using `except InvalidMetadataError` would help tracking it.